switchdev_poller: add new package (RFC)

luizluca opened the pull request:
This service monitors (each 3s) switchdev ports and brings down CPU
ports when all related non-CPU vlan ports are also down. Otherwise,
it brings the port up.

In order to hide CPU ports from netifd, when a device is brought down,
the device is renamed adding the suffix "_down".

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

Maintainer: me
Compile tested: OpenWrt 24d82c2e5c
Run tested: ar71xx (tl-wr1043nd)

Description:

neheb commented:
@luizluca please rebase against current master

diizzyy commented:
This should be in main repo imho

feckert commented:
You have explained what this does, but could you also explain for what I need this kind of service?

jow- commented:
@feckert - to example to restart the wan interface (DHCP negotation) when the wan cable is pulled and plugged back in (or when the router/modem/switch on the other side of the wan cable reboots).

Apparently this is a thing many people require. By default, Linux / netifd will not see a carrier loss on ethX.Y switch interfaces when the external ports go down because the connection between the internal switch IC and the CPU is always up.

feckert commented:
@jow thanks for your explanation
@luizluca If so could we rename them from switchdev_poller to switchdev-poller?
And could we add some debug output to syslog in /usr/sbin/switchdev_poller on start and stop?

luizluca commented:
> @luizluca If so could we rename them from switchdev_poller to switchdev-poller?
> And could we add some debug output to syslog in /usr/sbin/switchdev_poller on start and stop?

Sure. I'll do it in the following days.

luizluca commented:
@jow- @feckert , would it be another way to set a device down (and keep it down) instead of renaming the interface? netifd brings it up a little after I do "ip link set dev ${device} down". I'm currently renaming it to avoid netifd bringing it up again. However, it looks like a hack.

luizluca commented:
I just force-pushed a new version. Changes since first PR:

- renamed to switchdev-poller
- added disabled stderr parameter for init.d script (with a hint to use it for debug)
- moved script from /usr/sbin to /usr/lib/switchdev-poller (avoid $PATH pollution)
- added msg that informs when poller started/stopped and switch ports controls which devices:
```
Sun Apr 14 00:30:14 2019 daemon.info switchdev-poller[4131]: Monitoring switch0 (ports 2,3,4,5) for eth0.1
Sun Apr 14 00:30:14 2019 daemon.info switchdev-poller[4131]: Monitoring switch0 (ports 1) for eth0.2
Sun Apr 14 00:30:14 2019 daemon.info switchdev-poller[4131]: Monitoring switch0 (ports 2,3,4,5) for eth0.3
Sun Apr 14 00:30:14 2019 daemon.info switchdev-poller[4131]: Starting poller
Sun Apr 14 00:30:31 2019 daemon.info switchdev-poller[4131]: Stopped poller
```
@diizzyy, as I have commit access only to packages, it is easier for me to maintain it here. However, if any main repo committers wishes to move it to core, it's OK for me.

neheb commented:
Script should be ran through shellcheck IMO.

BKPepe commented:
@luizluca is this ready? Would you please look at neheb comment?

luizluca commented:
Shellchecked, but ignored non-POSIX usage of string replacement and indexing (as ash does support it).
Shellcheck helped uncover a bug where config_get had args switched. It was working because it reused the value defined previously. It would affect devices with two internal switches (if they exist).

Tested and ready for merging.

neheb commented on net/switchdev-poller/Makefile:
```diff
@@ -0,0 +1,42 @@
+   CATEGORY:=Network
+   DEPENDS:=+swconfig
+   TITLE:=Poll switchdev port to bring CPU ports up/down
+   MAINTAINER:=Luiz Angelo Daros de Luca <luizluca@gmail.com>
```
Unnecessary since PKG_MAIINTAINER is defined.

neheb commented on net/switchdev-poller/Makefile:
```diff
@@ -0,0 +1,42 @@
+PKG_RELEASE:=1
+PKG_MAINTAINER:=Luiz Angelo Daros de Luca <luizluca@gmail.com>
+PKG_LICENSE:=GPLv2
+PKG_ARCH:=all
```
This is incorrect and does nothing. The proper one is the one later.

neheb commented:
I usually get rid of those warnings by passing shellcheck -s dash as dash is similar enough to busybox ash.

luizluca commented:
> I usually get rid of those warnings by passing shellcheck -s dash as dash is similar enough to busybox ash.

Thanks but dash still does not support what I need:

```
ash$ a=123123123
ash$ echo ${a//3/_}
12_12_12_
```
```
dash$ a=123123123
dash$ echo ${a//3/_}
dash: 15: Bad substitution
```

I removed the unnecessary lines in Makefile.

BKPepe commented on net/switchdev-poller/Makefile:
```diff
@@ -0,0 +1,40 @@
+PKG_VERSION:=1.0.1
+PKG_RELEASE:=1
+PKG_MAINTAINER:=Luiz Angelo Daros de Luca <luizluca@gmail.com>
+PKG_LICENSE:=GPLv2
```
You are using outdated SPDX license. 
See https://spdx.org/licenses/

luizluca replied:
I'll use GPL-2.0-or-later

BKPepe commented on net/switchdev-poller/Makefile:
```diff
@@ -0,0 +1,40 @@
+define Package/switchdev-poller
+   SECTION:=net
+   CATEGORY:=Network
+   DEPENDS:=+swconfig
```
Shouldn't be depends under TITLE?

luizluca replied:
I wasn't aware that it should be a sequence. There are more pkgs just like that.
Anyway, I'll put it under.

luizluca commented:
As changes are mostly cosmetic, I'll a day or two if there is more suggestions before submitting a new merge.

luizluca commented:
It is done 41f240db6524e713161b2a97f0d5caae81b1fa10. Thanks.
